king_phisher/server/server_rpc.py

```diff
@@ -493,6 +494,63 @@ def rpc_database_insert_row(handler, session, table_name, keys, values):
 	session.commit()
 	return row.id
 
+@register_rpc('/db/table/insert/multi', database_access=True)
+def rpc_database_insert_row_multi(handler, session, table_name, keys, rows, deconflict_ids=False):
+	"""
+	Insert multiple new rows into the specified table. If deconflict_ids is true, new id values will be assigned
+	to merge the data into the database. This function will fail if contrasts for the table are not meet.
+
+	:param str table_name: The name of the database table to insert data into.
+	:param list keys: A list of keys of the column names in order of values to be inserted *values*
+	:param list(tuples) rows: A list of tuples, the tuple is the values to be inserted into the row.
+	:return: List of ids of the new rows inserted
+	:rtype: list
+	"""
+	inserted_ids = []
+	if not isinstance(keys, list):
+		keys = list(keys)
+	if not isinstance(rows, list):
+		rows = list(rows)
+
+	table = database_table_objects.get(table_name)
+	if not table:
+		raise errors.KingPhisherAPIError('failed to get table object for: {0}'.format(table_name))
+	for key in keys:
+		if key not in database_tables[table_name]:
+			raise errors.KingPhisherAPIError('column {0} is invalid for table {1}'.format(keys, table_name))
+
+	for row in rows:
+		if len(row) != len(keys):
+			errors.KingPhisherAPIError('row is not the same as number of values defined')
+		if deconflict_ids and 'id' in keys:
```

Comment:
Before entering this block, it might be easiest to convert row to a dictionary like: `row = dict(zip(keys, row))` so you can more easily check the `id` field exists, retrieve it's value etc. It would also allow you to later on replace the row object creation logic:

```
 +		table_row = table()
 +		for key, value in zip(keys, row):
 +			setattr(table_row, key, value)
```
with: `table_row = table(**row)`.